### DISCUSSION TITLE
Problem with Nuxt3 and Google Adsense

### DISCUSSION THREAD

**Discussion by yuriti** [nuxt/nuxt#16595] [Questions]
hello, I found SSR to work incorrectly with Google Adsense, it duplicates the components on the page, and in some places the entire page, has anyone come across this?

Disabling Google AdSense solves this problem but I need it.

**Comment by Flanker72** [comment]
It may appear when you have some hydration problems.
1. Check that you have EXACTLY one root element in your pages templates.
2. Try to wrap your AdSense code in ClientOnly.
3. Check that the script is loaded in the right place and only once.

**Reply by yuriti** [reply]
@Flanker72 Thank you for your help, but, unfortunately, everything is in order on all counts. Regarding your second point, I've tried without ClientOnly as well.

I would also like to clarify that everything worked well in Nuxt2.

Has anyone experienced this issue or are you using a different ad network?

**Reply by Flanker72** [reply]
I have an exactly the same issue with another component (not AdSense):  the component was duplicated on the page.

In my case this was because one of my components had nothing to render on the server (root with v-if leads to mismatch in server/client rendering). In the DevTools I had a warning related to incorrect hydration, and after making the component renderable at least with simple div instead of empty template the problem was solved.

**Reply by yuriti** [reply]
yes, this is a fairly common problem, but it didn't happen in my case. There are no warnings or errors, and duplication of components occurs only when Google Adsense is enabled, while nothing is displayed in the console. Unfortunately, I have already tried everything and made different approaches, but to no avail. Perhaps something else can be done.

I also want to add that duplication of components appears only on the first visit to the page, in case of transition from page to page, then everything is in order. Which is pretty weird.

**Comment by yuriti** [comment]
went deep into the problem and realized that if Google inserts ads in the site header (or in front of the component), then the nuxt breaks and duplicates the blocks. 

can nuxt with ssr enabled lose the block index (due to offset) and re-render it?

**Comment by mirumirumi** [comment]
Hello!
I spent three full days on this problem and was able to solve it in the following way.

```html
<template>
  <Head>
    <Script defer crossorigin="anonymous" src="https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js?client=ca-pub-xxxxxxxxxx"></Script>
  </Head>
  <div class="ad">
    <div ref="adins"></div>
    <div ref="adpush"></div>
  </div>
</template>

<script setup lang="ts">
const adins = ref()
const adpush = ref()

onMounted(() => {
  const ins = document.createElement("ins") as HTMLElement
  ins.classList.value = "adsbygoogle"
  ins.style.display = "inline-block"
  ins.setAttribute("data-ad-client", "ca-pub-xxxxxxxxxxxxxx")
  ins.setAttribute("data-ad-slot", "xxxxxxxxxxxxxxx")
  ins.setAttribute("data-ad-format", "rectangle")
  adins.value.appendChild(ins)
  
  const src = document.createElement("script") as HTMLScriptElement
  src.text = "(adsbygoogle = window.adsbygoogle || []).push({});"
  adpush.value.appendChild(src)
})
</script>
```

The points are:

- The "pagead2.googlesyndication.com~" script is loaded with the `<Head>` tag (async, defer, or none is OK)
- The body of the ad in the `<ins>` tag should also be written in DOM additions, not markup.
- The `adsbygoogle.push()` should be written by adding DOM as well.

The details are documented in [this issue](https://github.com/mirumirumi/mirumi-tech-frontend/issues/46) (sorry it's all in Japanese).

Hope this helps :)

**Comment by TheAlexLichter** [comment]
Does the problem also occur with the [adsense module](https://github.com/nuxt-community/google-adsense-module#readme)?
